core/src/main/java/io/atomix/core/map/impl/ConsistentMapService.java

```diff
@@ -153,7 +153,14 @@ public void restore(BufferInput<?> reader) {
     map = reader.readObject(serializer()::decode);
     activeTransactions = reader.readObject(serializer()::decode);
     currentVersion = reader.readLong();
-    map.forEach(this::scheduleTtl);
+    map.forEach((key, value) -> {
+      if (value.ttl() > 0) {
+        value.timer = scheduler().schedule(Duration.ofMillis(value.ttl() - (wallClock().time().unixTimestamp() - value.created())), () -> {
```

Comment:
Do we protect against negative milliseconds?

Reply:
I thought about that when I wrote this: the correctness of Raft and the state machine should guarantee that a negative duration is impossible. For that reason, we should rely on a negative TTL causing an exception, indicating an inconsistency in Raft or the state machine implementation.

On a related node, the scheduler needs the appropriate negative duration checks.

Reply:
#385

Reply:
Ok so the correctness of Raft and the state machine should avoid the situation of trying to schedule expiration of an expired entry on restoration..?

Reply:
Correct. A snapshot should always be taken and installed at the same point in logical time, therefore it should not be possible for an entry that should have expired to be present in the snapshot. The `wallClock()` is actually a deterministic approximation of wall clock time that’s based on times in the Raft log.